contrail-controller/hooks/contrail_controller_utils.py

```diff
@@ -179,3 +185,66 @@ def update_charm_status():
         docker_utils.compose_run(REDIS_CONFIGS_PATH + "/docker-compose.yaml")
 
     common_utils.update_services_status(SERVICES)
+
+
+def update_hosts_file(hosts_map):
+    """Update /etc/hosts file with cluster names and IPs.
+
+    RabbitMQ requires NODE names in a cluster to be resolvable.
+    https://www.rabbitmq.com/clustering.html#issues-hostname
+
+    In a multi-homed host scenario cluster IPs may have FQDNs structured
+    as interface_name.host_name.domain which will result in an issue if
+    a short name is derived from an FQDN by taking its first part.
+    See https://github.com/Juniper/contrail-charms/issues/50
+
+    This function updates /etc/hosts file with resolutions for IP -> hostname
+    lookups.
+    """
+    with open(HOSTS_FILE, 'r') as hosts:
+        lines = hosts.readlines()
+
+    log("Updating hosts file with: %s (current: %s)" % (hosts_map, lines),
+        level=INFO)
+
+    newlines = []
+    for ip, hostname in hosts_map.items():
+        if not ip or not hostname:
```

Comment:
is it realistic that ip or hostname can be empty?

Reply:
Probably not, unless a caller does it by accident. We could raise an exception here or just skip as it is done here.

The code of this function was literally copied from here: https://github.com/openstack/charm-rabbitmq-server/blob/stable/18.11/hooks/rabbit_utils.py#L663-L700